src/Modules/ModulesGraded.jl

```diff
@@ -1,5 +1,696 @@
+and standard degrees, that is the standard unit vectors have degree 0.
+
+The string `name` specifies how the basis vectors are printed. 
+
+# Examples
+```jldoctest
+julia> R, (x,y) = grade(polynomial_ring(QQ, ["x", "y"])[1])
+(Multivariate Polynomial Ring in x, y over Rational Field graded by 
+  x -> [1]
+  y -> [1], MPolyDecRingElem{QQFieldElem, QQMPolyRingElem}[x, y])
+
+julia> graded_free_module(R,3)
+Graded free module R^3([0]) of rank 3 over R
+
+```
+"""
+function graded_free_module(
+  R::Ring,
+  p::Int,
+  name::String="e",
+  d::Vector{GrpAbFinGenElem}=[grading_group(R)[0] for i in 1:p],
+)
+  @assert is_graded(R)
+  M = FreeMod(R, p, name)
+  M = grade(M, d)
+  return M
+end
+
+function graded_free_module(
+  R::Ring, d::Vector{GrpAbFinGenElem}=[grading_group(R)[0] for i in 1:p]
+)
+  return graded_free_module(R, length(d), "e", d)
+end
+
+function graded_free_module(R::Ring, W::Vector{<:IntegerUnion})
+  @assert is_graded(R)
+  A = grading_group(R)
+  return graded_free_module(R, length(W), "e", [W[i] * A[1] for i in 1:length(W)])
+end
+
+# TODO: move
+function is_graded(R::MPolyRing)
+  return false
+end
+
+function grade(M::FreeMod, d::Vector{GrpAbFinGenElem})
+  @assert length(d) == ngens(M)
+  @assert is_graded(base_ring(M))
+  R = base_ring(M)
+  N = free_module(R, length(d))
+  N.d = d
+  return N
+end
+
+function grading_group(M::FreeMod)
+  return grading_group(base_ring(M))
+end
+
+function grade(M::FreeMod, W::Vector{<:IntegerUnion})
+  @assert length(W) == ngens(M)
+  R = base_ring(M)
+  @assert is_graded(R)
+  A = grading_group(R)
+  N = free_module(R, length(W))
+  N.d = [W[i] * A[1] for i in 1:length(W)]
+  return N
+end
+
+function set_grading!(M::FreeMod, d::Vector{GrpAbFinGenElem})
+  @assert length(d) == ngens(M)
+  @assert is_graded(base_ring(M))
+  M.d = d
+end
+
+function set_grading!(M::FreeMod, W::Vector{<:IntegerUnion})
+  @assert length(W) == ngens(M)
+  R = base_ring(M)
+  @assert is_graded(R)
+  A = grading_group(R)
+  M.d = [W[i] * A[1] for i in 1:length(W)]
+end
+
+function degrees(M::FreeMod)
+  @assert is_graded(M)
+  return M.d
+end
+
+function degrees_of_generators(M::FreeMod)
+  return degrees(M)
+end
+
+
+function is_graded(M::FreeMod)
+  return isa(M.d, Vector{GrpAbFinGenElem})
+end
+
+###############################################################################
+# Graded Free Modules functions
+###############################################################################
+
+# for isomorphism of graded free modules, perhaps exists somewhere
+function compare_vectors_as_multisets(v1::Vector{T}, v2::Vector{T}) where {T<:Any}
+  v1_dict = Dict{T,Int}()
+  for entry in v1
+    if haskey(v1_dict, entry)
+      v1_dict[entry] += 1
+    else
+      v1_dict[entry] = 1
+    end
+  end
+  v2_dict = Dict{T,Int}()
+  for entry in v2
+    if haskey(v2_dict, entry)
+      v2_dict[entry] += 1
+    else
+      v2_dict[entry] = 1
+    end
+  end
+  return v1_dict == v2_dict
+end
+
+
+function swap!(A, i, j)
```

Comment:
Type annotations would help to make clear what this function does. I am guessing:
```suggestion
function swap!(A::Vector, i::Int, j::Int)
```

Reply:
Added type annotations for the functions finding the bijections between graded free module degrees.